### PR TITLE
Step4

### PR DESCRIPTION
**Pull request by ku-bo-mi**
## What
- Step4-1
- Step4-2
- Step4-3
- Step4-4
- Step4-5
- Step4-6

### Step4-2
スクリーンショット 2022-05-17 22 09 19

```
$ docker run -v $(pwd)/data/test_text.png:/tmp/img.png wakanapo/tesseract-ocr tesseract /tmp/img.png stdout -l eng
Build@Mercari Training Program 2022
This is @ku-bo-mi's build training repository.
```
### Step4-3
```
$ docker pull alpine
Using default tag: latest
latest: Pulling from library/alpine
df9b9388f04a: Pull complete 
Digest: sha256:4edbd2beb5f78b1014028f4fbb99f3237d9561100b6881aabbf5acce2c4f9454
Status: Downloaded newer image for alpine:latest
docker.io/library/alpine:latest

$ docker images
REPOSITORY               TAG       IMAGE ID       CREATED       SIZE
wakanapo/tesseract-ocr   latest    b13469b4999f   8 days ago    167MB
alpine                   latest    0ac33e5f5afa   6 weeks ago   5.57MB
```
### Step4-4
```
$ docker build -t build2022/app:latest python/
[+] Building 0.1s (5/5) FINISHED                                                                          
 => [internal] load build definition from Dockerfile                                                 0.0s
 => => transferring dockerfile: 36B                                                                  0.0s
 => [internal] load .dockerignore                                                                    0.0s
 => => transferring context: 2B                                                                      0.0s
 => [internal] load metadata for docker.io/library/alpine:latest                                     0.0s
 => CACHED [1/1] FROM docker.io/library/alpine                                                       0.0s
 => exporting to image                                                                               0.0s
 => => exporting layers                                                                              0.0s
 => => writing image sha256:a89e27d975c0864116673924636db1855be0e2a10510c7206dcbed3d7822d5cb         0.0s
 => => naming to docker.io/build2022/app:latest                                                      0.0s

Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them

$ docker images
REPOSITORY               TAG       IMAGE ID       CREATED       SIZE
wakanapo/tesseract-ocr   latest    b13469b4999f   8 days ago    167MB
alpine                   latest    0ac33e5f5afa   6 weeks ago   5.57MB
build2022/app            latest    a89e27d975c0   6 weeks ago   5.57MB
```

### Step4-5
```
mercari-build-training-2022 $ docker build -t build2022/app:latest python/
[+] Building 1.0s (5/5) FINISHED                                                                                        
 => [internal] load build definition from Dockerfile                                                               0.0s
 => => transferring dockerfile: 551B                                                                               0.0s
 => [internal] load .dockerignore                                                                                  0.0s
 => => transferring context: 2B                                                                                    0.0s
 => [internal] load metadata for docker.io/library/python:3                                                        0.9s
 => CACHED [1/1] FROM docker.io/library/python:3@sha256:567018293e51a89db96ce4c9679fdefc89b3d17a9fe9e94c0091b04ac  0.0s
 => exporting to image                                                                                             0.0s
 => => exporting layers                                                                                            0.0s
 => => writing image sha256:06c5cd4915d8bee2c3a840c3e1e302508c4ac1a557f4ab903102200f1cd80f2d                       0.0s
 => => naming to docker.io/build2022/app:latest                                                                    0.0s
Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them

$ docker run build2022/app
Python 3.10.4
```

### Step4-6
Build and run
```
(base) Nessie-5:mercari-build-training-2022 kbm$ docker build -t build2022/app:latest python/
[+] Building 1.0s (8/8) FINISHED                                                                                                                           
 => [internal] load build definition from Dockerfile                                                                                                  0.0s
 => => transferring dockerfile: 37B                                                                                                                   0.0s
 => [internal] load .dockerignore                                                                                                                     0.0s
 => => transferring context: 2B                                                                                                                       0.0s
 => [internal] load metadata for docker.io/library/python:3                                                                                           0.9s
 => [internal] load build context                                                                                                                     0.0s
 => => transferring context: 1.23kB                                                                                                                   0.0s
 => [1/3] FROM docker.io/library/python:3@sha256:567018293e51a89db96ce4c9679fdefc89b3d17a9fe9e94c0091b04ac5bb4e89                                     0.0s
 => CACHED [2/3] COPY . .                                                                                                                             0.0s
 => CACHED [3/3] RUN pip3 install --no-cache-dir --upgrade -r requirements.txt                                                                        0.0s
 => exporting to image                                                                                                                                0.0s
 => => exporting layers                                                                                                                               0.0s
 => => writing image sha256:33efee775bdd6ca8dbbac4a48a85945833530ec21ad6ec31cf0c38176650cee6                                                          0.0s
 => => naming to docker.io/build2022/app:latest                                                                                                       0.0s

Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them

(base) Nessie-5:mercari-build-training-2022 kbm$ docker run -d -p 9000:9000 build2022/app:latest
7cdaad2e9b95b88fbdea1c80bb0e615cee2ff87403ae24577e12d596b96e7aec

(base) Nessie-5:mercari-build-training-2022 kbm$ docker ps
CONTAINER ID   IMAGE                  COMMAND                  CREATED          STATUS          PORTS                    NAMES
7cdaad2e9b95   build2022/app:latest   "uvicorn main:app --…"   13 seconds ago   Up 13 seconds   0.0.0.0:9000->9000/tcp   brave_shtern
```
Works just like Step3
```
(base) Nessie-5:python kbm$ curl -X GET 'http://127.0.0.1:9000/items/10'
{"items":"[{'name': 'jacket', 'category': 'fashion', 'image': 'cbad5f2dbdd39f675ea9a3d499e116bff7cc86d51146d6a909ce9fd267e5f2a8.jpg'}]"}
```

### CODE REVIEW

File: python/main.py [hunk 1]
```diff
@@ -18,15 +21,62 @@
...
+Accepts the arguments as File.
+"""
 @app.post("/items")
-def add_item(name: str = Form(...)):
-    logger.info(f"Receive item: {name}")
+def add_item(name: bytes = File(...), category: bytes = File(...), image: bytes = File(...)):
+
+    # cast bytes to string
+    #name = name.decode('utf-8')
+    #category = category.decode('utf-8')
```
Comment (westcottian) [inline python/main.py]: Please remove these commented lines.

Comment (westcottian) [reply python/main.py]: Also, do we need this file to be check-in-ed?
[python/__pycache__/main.cpython-38.pyc](https://github.com/ku-bo-mi/mercari-build-training-2022/pull/3/files#diff-1010e044d2695a90779ad770a553cda3c4f31d0b602c81c565dc75d5c6f3984b)

Comment (ku-bo-mi) [reply python/main.py]: Thanks! I actually shouldn't have commented out. I put them back to code.

**Comment by CarnevaleMNO** [comment]
Docker and other tools build these container images incrementally, adding one layer on top of the other, starting from the top of the Dockerfile and adding any files created by each of the instructions of the Dockerfile.

Docker and similar tools also use an internal cache when building the image, if a file hasn't changed since the last time building the container image, then it will re-use the same layer created the last time, instead of copying the file again and creating a new layer from scratch.

Just avoiding the copy of files doesn't necessarily improve things too much, but because it used the cache for that step, it can use the cache for the next step. For example, it could use the cache for the instruction that installs dependencies with:


RUN pip install --no-cache-dir --upgrade -r /code/requirements.txt
The file with the package requirements won't change frequently. So, by copying only that file, Docker will be able to use the cache for that step.

And then, Docker will be able to use the cache for the next step that downloads and install those dependencies. And here's where we save a lot of time. ✨ ...and avoid boredom waiting. 😪😆

Downloading and installing the package dependencies could take minutes, but using the cache would take seconds at most.

And as you would be building the container image again and again during development to check that your code changes are working, there's a lot of accumulated time this would save.

Then, near the end of the Dockerfile, we copy all the code. As this is what changes most frequently, we put it near the end, because almost always, anything after this step will not be able to use the cache.


COPY ./app /code/app

**Comment by ku-bo-mi** [comment]
Thank you @CarnevaleMNO ! I changed the codes and now it builds much more quickly!

**Pull request merged**